### PR TITLE
🐛 [CORE-17299] Log all parameters. Added basic testing

### PR DESCRIPTION
**Pull request by lexasa**
Added logging of all parameters
Added basic testing
Cosmetics

### CODE REVIEW

File: packages/logger/jest.config.js [hunk 1]
```diff
@@ -7,25 +7,14 @@ module.exports = {
...
-    '/node_modules/',
-  ],
-  coveragePathIgnorePatterns: [
-    '/node_modules/',
-    'src/index.ts',
-    'src/config',
-  ],
+  testRegex: '/tests/.*\\.(ts|tsx|js)$',
+  testPathIgnorePatterns: ['/node_modules/', '/build/'],
+  coveragePathIgnorePatterns: ['/node_modules/', 'src/index.ts', 'src/config'],
```
Comment (knor-el-snor) [inline packages/logger/jest.config.js:19]: Thanks!

File: packages/logger/src/index.ts [hunk 2]
```diff
@@ -1,32 +1,42 @@
...
+const instance: Logger = createLogger({
   level: LOG_LEVEL,
-  format: winston.format.combine(
-    winston.format.colorize(),
-    winston.format.timestamp(),
-    winston.format.printf(({ level, message }) => `${formatLogLevel(level)}: ${message}`),
+  format: format.combine(
+    emojiLevel(),
+    format.colorize(),
+    format.timestamp({ alias: 'timestamp' }),
```
Comment (knor-el-snor) [inline packages/logger/src/index.ts:43]: I believe this could be a discussion point, we've added this coming from some best practices but our GCP logs and your local terminal (configurable) will already have a timestamp built-in by default. @gorjuspixels @nikolaev-k What would you do here, I remember we discussed this a loooong time ago?

File: packages/logger/src/tests/logger.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    expect(spyOnConsole).toHaveBeenCalledWith<[string]>(
+      expect.stringContaining('1970-01-01T00:00:00.000Z: message\n{"param":"test"}\n'), // cut off color string
+    );
+  });
+
+  it('debug off', () => {
+    logger.debug(message, params[0], params[1]);
+    expect(spyOnConsole).not.toBeCalledTimes(1);
+  });
+  it('error', () => {
```
Comment (knor-el-snor) [inline packages/logger/src/tests/logger.ts]: Please add newline between `if` cases such as described [here](https://coda.io/d/Development-Guidelines_d7DszrlIOzU/Coding_suXNI#_luxdJ).

File: packages/logger/package.json [hunk 4]
```diff
@@ -40,13 +40,18 @@
...
   },
   "devDependencies": {
     "@types/debug": "4.1.5",
+    "@types/faker": "5.1.2",
     "@types/jest": "26.0.7",
+    "@types/json-stringify-safe": "5.0.0",
+    "faker": "5.1.0",
     "husky": "4.2.5",
     "jest": "26.1.0",
+    "mockdate": "3.0.2",
```
Comment (knor-el-snor) [inline packages/logger/package.json]: You added this package but seems it isn't used anywhere? If so, please remove from the package.json and run bootstrap again so it's gone from our lockfile too

File: packages/logger/src/index.ts [hunk 5]
```diff
@@ -1,32 +1,42 @@
...
-    case 'info':
-      return `🍺 ${level}`;
-    case 'warn':
-      return `❗️ ${level}`;
-    case 'error':
-      return `🔥 ${level}`;
-    default:
-      return `🤷‍♂️ ${level}`;
-  }
+const LevelEmoji: Record<string, string> = {
```
Comment (knor-el-snor) [inline packages/logger/src/index.ts]: Please use `LEVEL_EMOJI` as referred [here](https://coda.io/d/Development-Guidelines_d7DszrlIOzU/Naming_suckZ#_lugiE) for these constants.

File: packages/logger/src/tests/logger.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
+import { NSlogger } from '..';
+import { lorem } from 'faker';
+
+const mockDate = new Date(0);
+jest.spyOn(global, 'Date').mockImplementation(() => (mockDate as unknown) as string); // TypeScript workaround
```
Comment (knor-el-snor) [inline packages/logger/src/tests/logger.ts]: Would it be possible to just use (beforeEach not required I believe)
```
  beforeEach(() => {
    Date.now = () => 1602254300000
  })
```

Not sure what Winston uses in the background but normally this should work, and doesn't give any TS issues

File: packages/logger/src/tests/logger.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+const OLD_ENV = process.env;
+
+describe('logger', () => {
+  const logger = NSlogger('test');
+
+  beforeEach(() => {
+    jest.resetModules();
+    jest.clearAllMocks();
+    process.env = { ...OLD_ENV };
```
Comment (knor-el-snor) [inline packages/logger/src/tests/logger.ts]: What is the goal of all this setup if you don't update the `process.env` variables? Also, if you'd wish to update those and set them for a specific test, I have a cleaner way of doing so instead of this `OLD_ENV` trickery ;-)

File: packages/logger/src/tests/logger.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+  it('debug off', () => {
+    logger.debug(message, params[0], params[1]);
+    expect(spyOnConsole).not.toBeCalledTimes(1);
+  });
+  it('error', () => {
+    logger.error(message, params[0], params[1]);
+    expect(spyOnConsole).toBeCalledTimes(1);
+  });
+  it('warn', () => {
```
Comment (knor-el-snor) [inline packages/logger/src/tests/logger.ts]: A test case should basically describe whatever the expect outcome will be and technically, even though not always applicable, make it so that all user stories are there just by reading the `it` blocks. We prefer to use the `Should x .... if ...` way of describing test cases in a full sentence. This is not needed for `describe` blocks. Most of these are describe in our [Development Guidelines](https://coda.io/d/Development-Guidelines_d7DszrlIOzU/Testing_suPTM#_luOL2). If they're not explicitly in there, definitely feel free to address or start a discussion about this so we can align across the team

For example in this case (not saying it's a perfect one, just for the gist of it):

```
it('Should return a warning log accepting multiple parameters')
```

File: packages/logger/src/tests/logger.ts [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  { [lorem.word()]: { [lorem.word()]: lorem.sentence() } },
+];
+
+const OLD_ENV = process.env;
+
+describe('logger', () => {
+  const logger = NSlogger('test');
+
+  beforeEach(() => {
+    jest.resetModules();
```
Comment (knor-el-snor) [inline packages/logger/src/tests/logger.ts]: Why is this needed? This would only be applicable if you were mocking a module with different behaviour within every test case, I don't this needed in this test file?

File: packages/logger/src/tests/logger.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
+import { NSlogger } from '..';
+import { lorem } from 'faker';
+
+const mockDate = new Date(0);
+jest.spyOn(global, 'Date').mockImplementation(() => (mockDate as unknown) as string); // TypeScript workaround
+const spyOnConsole = jest.spyOn((console as any)._stderr, 'write').mockImplementation();
```
Comment (knor-el-snor) [inline packages/logger/src/tests/logger.ts]: Please "Use verb to express the action which this function does and add spy suffix." as described in our [Development Guidelines](https://coda.io/d/Development-Guidelines_d7DszrlIOzU/Testing_suPTM#_lux6N). This would lead to `consoleSpy`

File: packages/logger/src/tests/logger.ts [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+  });
+  it('warn', () => {
+    logger.warn(message, params[0], params[1]);
+    expect(spyOnConsole).toBeCalledTimes(1);
+  });
+
+  it('real console output', () => {
+    spyOnConsole.mockRestore();
+    logger.warn(message, params[0], params[1]);
+  });
```
Comment (knor-el-snor) [inline packages/logger/src/tests/logger.ts]: Nothing actually gets tested here, there's no expect to see what the actual output is. What would be the goal in this case?

File: packages/logger/src/index.ts [hunk 12]
```diff
@@ -1,32 +1,42 @@
...
+  info: '🍺',
+  warn: '❗️',
+  error: '🔥',
+  default: '🤷‍♂️',
 };
 
-const instance: winston.Logger = winston.createLogger({
+// have to cast to any to access additional params value
+const splatIndex = Symbol.for('splat') as any;
```
Comment (knor-el-snor) [inline packages/logger/src/index.ts]: I see 2 possible ways of making this a bit more readable for people who didn't have to go through forums or Winston docs and still make this readable.
1. Add a JSDoc right above explaining what it does, why are you looking for that specific one. `// Winston adds all parameters in their 'splat' property...`
2. Make this into a function. 

```
/**
* Returns all available Winston input values.
* @param {object} info - Winston information when logging gets used.
*/
const getWinstonInputValues = <T>(info: TransformableInfo):T => Symbol.for('splat') as T;
```

Potentially even add this logic in the same function and rename it to `getWinstonParameters`.  
```
return paramsIn instanceof Array ? paramsIn : [paramsIn]).map((v: unknown) => EOL + stringify(v);

```

File: packages/logger/src/index.ts [hunk 13]
```diff
@@ -1,32 +1,42 @@
...
+
+const emojiLevel = format(
```
Comment (knor-el-snor) [inline packages/logger/src/index.ts]: I would rename this function to something that gives a better indication of what it returns. The current name would indicate it just returns the level emoji, which it also does, but it returns the entire stringified parameters too. I would suggest creating the following:

```
getEmojiLevel = ({ level }: TransformableInfo) => LEVEL_EMOJI[level] || LEVEL_EMOJI.default;
```

Combining this with the comment above to add another function, I would just call it `customFormat(...)` or something like that as a function name 🤷‍♂️

File: packages/logger/tsconfig.json [hunk 14]
```diff
@@ -12,21 +13,14 @@
     "declaration": true,
 
     "noImplicitAny": false,
-    "allowSyntheticDefaultImports": false,
+    "allowSyntheticDefaultImports": true,
```
Comment (knor-el-snor) [inline packages/logger/tsconfig.json:16]: This used to give a lot of problems combined with jest, as where the project would build but then jest wouldn't run tests because of an import and vice-versa. It's fixed now in `ts-jest` so great you noticed to turn this on 👌

File: packages/logger/tsconfig.json [hunk 15]
```diff
@@ -12,21 +13,14 @@
...
     "removeComments": true,
     "noUnusedLocals": true,
-    "noUnusedParameters": true,
-
+    "noUnusedParameters": false,
```
Comment (knor-el-snor) [inline packages/logger/tsconfig.json]: I don't think this should be turned off. If you wish to have a parameter in these you can prefix it with `_` as what Typescript offers to go around this issue in those edge cases when needed. It's a great tool to see which parameters have been forgotten even though eslint checks it more strict in our current eslint rules I believe

**Review by knor-el-snor — CHANGES_REQUESTED** [review]
I know this is an open-source project that kind of got created as a side thingy but I know it's mostly just being used and definitely maintained by Bonsai so I think it would be great to keep it kind of in line with the development guidelines for internal projects. This will make it easier to switch between closed & open source projects for Bonsai too I believe. Probably we'll open-source our guidelines (since they're not really a secret) somewhere so we can add these in the How To Contribute for these open-source projects

File: packages/logger/jest.config.js (hunk 1 above)
Comment (lexasa) [reply packages/logger/jest.config.js:19]: it was basically auto-formatted :)

File: packages/logger/src/index.ts (hunk 2 above)
Comment (lexasa) [reply packages/logger/src/index.ts:43]: Yes, plain text format is good only for debug purposes. GCP doesn't render colorization. May be it's good idea to have different logging setup for dev/prod environment.

File: packages/logger/src/tests/logger.ts (hunk 3 above)
Comment (lexasa) [reply packages/logger/src/tests/logger.ts]: NP, I think all code formatting requirements must be in eslint config.

File: packages/logger/package.json (hunk 4 above)
Comment (lexasa) [reply packages/logger/package.json]: yep, played around with mocking and didn't remove it. good catch

File: packages/logger/src/tests/logger.ts (hunk 7 above)
Comment (lexasa) [reply packages/logger/src/tests/logger.ts]: Leftover from unwritten debug test case :) I'll be glad to knowhow do you do it

File: packages/logger/src/tests/logger.ts (hunk 8 above)
Comment (lexasa) [reply packages/logger/src/tests/logger.ts]: got it

File: packages/logger/src/tests/logger.ts (hunk 9 above)
Comment (lexasa) [reply packages/logger/src/tests/logger.ts]: old habit to clean everything. will remove

File: packages/logger/src/tests/logger.ts (hunk 11 above)
Comment (lexasa) [reply packages/logger/src/tests/logger.ts]: you're right. it's pure debug thing for human.

File: packages/logger/src/tests/logger.ts (hunk 6 above)
Comment (lexasa) [reply packages/logger/src/tests/logger.ts]: I would prefer not to mess with prototypes and mock anything outside of Jest scope.

Comment (lexasa) [reply packages/logger/src/tests/logger.ts]: Date.now = () => ... doesn't work

File: packages/logger/src/tests/logger.ts (hunk 3 above)
Comment (knor-el-snor) [reply packages/logger/src/tests/logger.ts]: @lexasa Yeah you're right and I think some of them are already in there but this open-source project still uses the old rules unfortunately. And having newlines between `it` cases is currently in the manual guidelines, but is planned to add into our `eslint-config`. And then open-source our eslint-config so we can use our linting rules also in open-source or other packages more easily. At least that's what I would like to see. We aren't there **yet**

File: packages/logger/src/index.ts [hunk 16]
```diff
@@ -1,32 +1,50 @@
...
+
+const instance: Logger = createLogger({
   level: LOG_LEVEL,
-  format: winston.format.combine(
-    winston.format.colorize(),
-    winston.format.timestamp(),
-    winston.format.printf(({ level, message }) => `${formatLogLevel(level)}: ${message}`),
+  format: format.combine(
+    emojiLevelFormat(),
+    paramsFormat(),
```
Comment (knor-el-snor) [inline packages/logger/src/index.ts:41]: Nice, splitting them up makes it more modular, I like it @lexasa!

File: packages/logger/src/tests/logger.ts [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+});
+
+it('Should output formatted error message to console', () => {
+  logger.error(message, params[0], params[1]);
+  expect(consoleSpy).toBeCalledTimes(1);
+});
+it('Should output formatted warn message to console', () => {
+  logger.warn(message, params[0], params[1]);
+  expect(consoleSpy).toBeCalledTimes(1);
+});
```
Comment (knor-el-snor) [inline packages/logger/src/tests/logger.ts]: I'll create a story to add the automatic newline between `it` cases in our `eslint-config` but for now it's still a manual check 😅

```suggestion
it('Should output formatted error message to console', () => {
  logger.error(message, params[0], params[1]);
  expect(consoleSpy).toBeCalledTimes(1);
});

it('Should output formatted warn message to console', () => {
  logger.warn(message, params[0], params[1]);
  expect(consoleSpy).toBeCalledTimes(1);
});
```

File: packages/logger/src/tests/logger.ts [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+const consoleSpy = jest.spyOn((console as any)._stderr, 'write').mockImplementation();
+
+const message = lorem.sentence();
+const params: Record<string, unknown>[] = [
+  { [lorem.word()]: { [lorem.word()]: lorem.sentence() }, [lorem.word()]: lorem.sentence() },
+  { [lorem.word()]: { [lorem.word()]: lorem.sentence() } },
+];
+
+const logger = NSlogger('test');
+
```
Comment (knor-el-snor) [inline packages/logger/src/tests/logger.ts:15]: Am I missing something but wasn't there a main `describe` block? If not could you please add it to wrap our hooks & `it` cases here. Can be just `Logger basic tests` or whatever you'd like to call them

File: packages/logger/src/tests/logger.ts (hunk 6 above)
Comment (knor-el-snor) [reply packages/logger/src/tests/logger.ts]: @lexasa I think I've implemented a different solution in the past without Typescript issues but not worth looking for it here right now so makes sense to stay off the `Date.now` 👌

File: packages/logger/src/tests/logger.ts [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+beforeEach(() => {
+  jest.clearAllMocks();
+});
+
+afterAll(() => {
+  jest.restoreAllMocks();
+});
```
Comment (knor-el-snor) [inline packages/logger/src/tests/logger.ts]: Still wondering why this would still be here? Every test suite is fully isolated within jest so shouldn't be clean between test files unless I'm mistaken. Why would you need to restore these afterAll?

**Review by knor-el-snor — CHANGES_REQUESTED** [review]
A couple super small things and we're ready to merge!
Please make sure to apply the 👍 on PR comments making it easier for a rereview to check what you recognize that you've seen and addressed to make sure you didn't miss anything. It's described [here ](https://coda.io/d/Development-Guidelines_d7DszrlIOzU/Pull-Requests-PR_sukry#_luabE) if you want more information

**Review by knor-el-snor — APPROVED** [review]

**Comment by knor-el-snor** [comment]
Please still update the `CHANGELOG.md`, forgot about that!
